File: delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* travelandchange(TreeNode* root, int target)
    {
        if(root == NULL)
        {
            return NULL;
        }
        root->left = travelandchange(root->left,target);
        root->right = travelandchange(root->right,target);
        //postorder - check node's value and delete if == target
        if(root->val == target && root->left == NULL && root->right == NULL)
        {
            return NULL;
        }
        return root;
    }
    
    TreeNode* removeLeafNodes(TreeNode* root, int target) {
        TreeNode* node = root;
        TreeNode* ans = travelandchange(node,target);
        return ans;
    }
};

File: delete-leaves-with-a-given-value/delete-leaves-with-a-given-value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    public TreeNode travel(TreeNode node,int target)
    {
        if(node == null)
        {
            return null;
        }
        TreeNode rleft = travel(node.left,target);
        TreeNode rright = travel(node.right,target);
        node.left = rleft;
        node.right = rright;
        if(rleft == null && rright == null && node.val == target)
        {
            return null;
        }
        return node;
    }
    
    public TreeNode removeLeafNodes(TreeNode root, int target) {
        TreeNode node = root;
        root = travel(node,target);
        return root;
    }
}